source/http_server.c

```diff
@@ -0,0 +1,228 @@
+    } else {
+        aws_mem_release(py_server->allocator, py_server);
+    }
+}
+
+static void s_on_incoming_connection(
+    struct aws_http_server *server,
+    struct aws_http_connection *connection,
+    int error_code,
+    void *user_data) {
+    (void)server;
+    (void)connection;
+    (void)error_code;
+    (void)user_data;
+    /* fake one */
+}
+
+PyObject *aws_py_http_server_create(PyObject *self, PyObject *args) {
+    (void)self;
+
+    struct py_http_server *py_server = NULL;
+    struct aws_allocator *allocator = aws_crt_python_get_allocator();
+
+    PyObject *bootstrap_capsule = NULL;
+    PyObject *on_incoming_connection = NULL;
+    PyObject *on_destroy_complete = NULL;
+    const char *host_name = NULL;
+    Py_ssize_t host_name_len = 0;
+    uint16_t port_number = 0;
+    PyObject *py_socket_options = NULL;
+    PyObject *tls_conn_options_capsule = NULL;
+
+    if (!PyArg_ParseTuple(
+            args,
+            "OOOs#HOO",
+            &bootstrap_capsule,
+            &on_incoming_connection,
+            &on_destroy_complete,
+            &host_name,
+            &host_name_len,
+            &port_number,
+            &py_socket_options,
+            &tls_conn_options_capsule)) {
+        PyErr_SetNone(PyExc_ValueError);
+        goto error;
+    }
+
+    if (host_name_len >= (signed int)AWS_ADDRESS_MAX_LEN || host_name_len == 0) {
+        PyErr_SetString(PyExc_ValueError, "host_name is not valid");
+        goto error;
+    }
+
+    if (py_socket_options == Py_None) {
+        PyErr_SetString(PyExc_ValueError, "socket_options is a required argument");
+        goto error;
+    }
+
+    if (!PyCallable_Check(on_incoming_connection)) {
+        PyErr_SetString(PyExc_ValueError, "on_incoming_connection callback is required");
+        goto error;
+    }
+
+    if (!PyCallable_Check(on_destroy_complete)) {
+        PyErr_SetString(PyExc_TypeError, "on_destroy_complete is invalid");
+        goto error;
+    }
+
+    struct aws_server_bootstrap *bootstrap = PyCapsule_GetPointer(bootstrap_capsule, s_capsule_name_server_bootstrap);
+    if (!bootstrap) {
+        PyErr_SetString(PyExc_ValueError, "the bootstrap capsule has an invalid pointer");
+        goto error;
+    }
+
+    struct aws_tls_connection_options *connection_options = NULL;
+
+    if (tls_conn_options_capsule != Py_None) {
+        connection_options = PyCapsule_GetPointer(tls_conn_options_capsule, s_capsule_name_tls_conn_options);
+    }
+
+    py_server = aws_mem_calloc(allocator, 1, sizeof(struct py_http_server));
+    if (!py_server) {
+        PyErr_SetAwsLastError();
+        goto error;
+    }
+
+    struct aws_socket_options socket_options;
+    AWS_ZERO_STRUCT(socket_options);
+    if (!aws_socket_options_init_from_py(&socket_options, py_socket_options)) {
+        goto error;
+    }
+
+    Py_INCREF(on_incoming_connection);
+    py_server->on_incoming_connection = on_incoming_connection;
+
+    Py_INCREF(on_destroy_complete);
```

Comment:
you might want to wait until you're super sure that everthing has succeeded before doing Py_INCREF.

if you do it too early in the function, and errors might occur later, you have to worry whether or not to DECREF the object